### DISCUSSION TITLE
55,000 static pages? Is this crazy?

### DISCUSSION THREAD

**Discussion by baptisteArno** [vercel/next.js#20832] [Help]
Currently our build generates 1855 static page. But it takes a while (~5 minutes here):
```
17:02:07.231 | info  - Generating static pages (0/1855)
17:04:12.332 | info  - Generating static pages (463/1855)
17:05:58.974 | info  - Generating static pages (927/1855)
17:07:39.023 | info  - Generating static pages (1391/1855)
17:09:14.341 | info  - Generating static pages (1855/1855)
```

It's calling a free Heroku dyno. Upgrading would make this faster?

Also, we're planning on adding a large amount of data that will result in at least 55,000 static pages. Should I drop Static Generation?

**Comment by paulogdm** [comment]
I would argue that using ISG instead ([reference](https://nextjs.org/docs/basic-features/data-fetching#incremental-static-regeneration)) would be better since the content is generated on-demand, and your build can finish quicker.

At scale, generating tons of static pages can work, but should you do it? The answer depends on a few factors, but perhaps you can generate only 5K pages (most accessed) and leave the rest to be done when requests are received.

Upgrading a dyno should boost your performance since free instances tend to be the bare minimum. You should never use a free dyno in production since problems on that front will cost you down the line.

**Reply by baptisteArno** [reply]
Thank you for taking the time @paulogdm 

> I would argue that using ISG instead ([reference](https://nextjs.org/docs/basic-features/data-fetching#incremental-static-regeneration)) would be better since the content is generated on-demand, and your build can finish quicker.

Oh yes I already use ISG but the build process still generates static pages for the first time even if it has "revalidate" option. Or am I missing something?

> you can generate only 5K pages (most accessed) and leave the rest to be done when requests are received.

How can I generate statically only a subset of pages?

**Reply by paulogdm** [reply]
I am moving this discussion to the Next.js repository.

You can return a subset of page paths on `getStaticPaths`. That way only those pages will be built while everything else will be generated "on demand".

**Comment by michrome** [comment]
Re Heroku, you need to be mindful that the build time is limited to 15 mins and, if you're not on performance Dynos, that build time will ["vary wildly" according to Heroku](https://devcenter.heroku.com/articles/slug-compiler#time-limit). If you're currently on free, a Performance dyno will look expensive, starting at [$250/month for a single M dyno](https://www.heroku.com/pricing).

I'd agree with @paulogdm that using `getStaticPaths` to limit SSG builds is a good approach. If you do that and still have a build time over 10 mins, you might want to look at using Vercel, which gives you a [45 mins build window](https://vercel.com/docs/platform/limits).

**Reply by baptisteArno** [reply]
I'm currently deploying on Vercel but my CMS (Strapi) is deployed on Heroku. So to generate the pages, it calls the API hosted in a Free Dyno. If I upgrade my Heroku plan, will it speed up Vercel build?

**Reply by michrome** [reply]
I understand now. If you're building on Vercel then you should get a good amount of parallelism. I don't know how many cores their build servers have but [Next will use them all](https://github.com/vercel/next.js/issues/12360#issuecomment-622424524)!

Your build times _could_ be reduced if either of the following is true:

- Your CMS API responses are slow because they are CPU bound i.e. you're doing some computation in the dyno. If this is true, then a performance dyno would help (vertical scaling). I suspect it's probably not true, though<sup>*</sup>.
- Your CMS API responses are slow because you're sending too many concurrent requests to the dyno. If this is true, then adding standard dynos would help (horizontal scaling).

While you could do lots of profiling, I'd be tempted to just try it: upgrade the dynos, run a few tests, downgrade the dynos. You'll only spend a couple of dollars and you'll know for sure.


<sup>*</sup> If the Heroku side is slow, my prime suspect would be the database, not the CPU: [the free/hobby Postgres plans have 0 bytes RAM](https://elements.heroku.com/addons/heroku-postgresql) for caching and are limited to 20 connections.
